### ISSUE TITLE
Better support for multi architectures (32 and 64 bits)

### ISSUE THREAD

**Issue by wismill** [commercialhaskell/stack#327]
Hi, I am using Windows 7 *64 bits* and it is now quite easy to install GHC in both 32 and 64 bits by using the following commands:

- `./stack setup --arch i386`
- `./stack setup --arch x86_64`.

Proposition 1: **support a list of architecture rather than just one**:

-  `./stack setup --arch i386,x86_64`
- In `stack.yaml`configuration file, arch option should manage *a list* of architectures (separated with commas).

I am using the *64 bits version of GHC by default* (added to my `PATH` environment variable).

I tried to build my project and I encounter some strange issues:

- `./stack build` seems to default to i386 but it builds x86_64. In fact, it creates subfolders `i386-windows` in `.stack-work/dist` and `.stack-work/install`; but in `.stack-work\install\i386-windows\nightly-2015-06-16\7.10.1\bin` I got a 64 bits executable. Maybe because stack uses the GHC referenced in `PATH`?
- `./stack build --arch XXX` accept whatever value for XXX, creates `XXX-windows` in `.stack-work/dist` and `.stack-work/install` but still builds 64 bits, even with the valid `i386` value for XXX.

Proposition 2: **`stack build` should be consistent between the architecture it targets (default, `--arch` option or `stack.yaml`), the available GHC versions installed via stack and GHC referenced in the `PATH` environment variable.**

Proposition 3: **`stack build` should build by default all the architectures defined in `stack.yaml`, except if the option `arch` is supplied.**

(possible linked issue: #132 ?)

**Comment by snoyberg** [comment]
I don't understand what advantage this would have vs just building twice, once with each architecture. I'm also not sure what proposition 2 is saying, but I _do_ believe that #132 may be related.

Can you clarify: why isn't `stack build --arch i386 && stack build --arch x86_64` sufficient? There are dozens if not hundreds of design choices that would need to be made if we were to perform completely different builds with a single command, and I'd much rather avoid that if there's not serious benefit.

**Comment by wismill** [comment]
Well, `stack build --arch i386 && stack build --arch x86_64` is sufficient to do the job. Proposition 3 is more about convenience, as proposition 1. But if you want stack to really facilitate the dev's job, they make a lot of sense.

Proposition 2 deals with fixing a bug.

**Comment by snoyberg** [comment]
Is "proposition 2" any different from #132? I'm having a hard time following exactly what bug is being reported.

**Comment by snoyberg** [comment]
FYI, I went ahead and implemented #132. If it's easier to test out a new version that review the intended implementation there, you could try out the latest master and see if it resolves your "proposition 2". I tested on Windows, and stack now correctly ignores a system GHC if it has the wrong architecture.

**Comment by wismill** [comment]
Let me clarify:

- `./stack build` compiles to 64 bits but write in `i386-windows` folders instead of `x86_64`
- `./stack build --arch i386` ends up with the previous issue
- `./stack build --arch XXX` accept any value for XXX, even illegal ones, and compiles to 64 bits in folders `XXX-windows`.

I have change `PATH` to reference GHC 32 bits instead of GHC 64 bits and I got `./stack build` working well for i386.

So to summarise:

- option `arch` in `./stack build --arch XXX` only affects the target directory, not the compiled files.
- stack uses only the GHC version of `PATH`, where it should use the version adapted to the target.

I will try the latest master soon.

**Comment by wismill** [comment]
Great, it seems to work with revision 8b8eeea0d0c566de02a4577256d5544995954923!

But I still get lots of warnings:

```
ghc: warning: _tzset from msvcrt is linked instead of __imp__tzset

test-stack-0.1.0.0: build
Building test-stack-0.1.0.0...
Preprocessing executable 'test-stack' for test-stack-0.1.0.0...
[1 of 1] Compiling Main             ( Main.hs, .stack-work\dist\x86_64-windows\Cabal-1.22.2.0\build\test-stack\test-stack-tmp\Main.o )
Linking .stack-work\dist\x86_64-windows\Cabal-1.22.2.0\build\test-stack\test-stack.exe ...
test-stack-0.1.0.0: install
Installing executable(s) in
C:\Users\qrn\Downloads\test-stack\.stack-work\install\x86_64-windows\nightly-2015-06-15\7.10.1\bin
ghc-pkg --no-user-package-db --package-db=C:\Users\qrn\AppData\Roaming\stack\snapshots\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\ --package-db=C:\Users\qrn\Downloads\test-stack\.stack-work\install\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\ describe test-stack:
WARNING: cache is out of date: C:\Users\qrn\Downloads\test-stack\.stack-work\install\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\package.cache
ghc will see an old view of this package db. Use 'ghc-pkg recache' to fix.
WARNING: cache is out of date: C:\Users\qrn\AppData\Roaming\stack\snapshots\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\package.cache
ghc will see an old view of this package db. Use 'ghc-pkg recache' to fix.
WARNING: cache is out of date: C:/Users/qrn/AppData/Local/Programs/stack/x86_64-windows/ghc-7.10.1\lib\package.conf.d\package.cache
ghc will see an old view of this package db. Use 'ghc-pkg recache' to fix.
ghc-pkg.exe: cannot find package test-stack

ghc-pkg --no-user-package-db --package-db=C:\Users\qrn\AppData\Roaming\stack\snapshots\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\ --package-db=C:\Users\qrn\Downloads\test-stack\.stack-work\install\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\ describe test-stack:
WARNING: cache is out of date: C:\Users\qrn\Downloads\test-stack\.stack-work\install\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\package.cache
ghc will see an old view of this package db. Use 'ghc-pkg recache' to fix.
WARNING: cache is out of date: C:\Users\qrn\AppData\Roaming\stack\snapshots\x86_64-windows\nightly-2015-06-15\7.10.1\pkgdb\package.cache
ghc will see an old view of this package db. Use 'ghc-pkg recache' to fix.
WARNING: cache is out of date: C:/Users/qrn/AppData/Local/Programs/stack/x86_64-windows/ghc-7.10.1\lib\package.conf.d\package.cache
ghc will see an old view of this package db. Use 'ghc-pkg recache' to fix.
ghc-pkg.exe: cannot find package test-stack
```

**Comment by snoyberg** [comment]
Let's track the issue of the warning output separately (I think it's a quick fix, I just pinged @chrisdone about it). Beyond that, is everything working as expected at this point for the multi-arch case?

**Comment by wismill** [comment]
Hi, working with revision 02051f0c4b8c69dd190ab64c04ff94ca8d8f6e81 I got the issue solved: many thanks! I tested it at work (Windows 7) and at home (Windows 8). The following three commands work as expected:

- `stack build` : default to architecture: i386 (see my comment below).
- `stack build --arch *incorrectArch*` : error as expected
- `stack build --arch i386`
- `stack build --arch x86_64`

Hereinafter the log of the build:

```
ghc: warning: _tzset from msvcrt is linked instead of __imp__tzset

dummy-project-0.1.0.0: build
Building dummy-project-0.1.0.0...
Preprocessing executable 'dummy-project' for dummy-project-0.1.0.0...
[1 of 1] Compiling Main             ( src\Main.hs, .stack-work\dist\x86_64-windows\Cabal-1.22.2.0\build\dummy-project\dummy-project-tmp\Main.o )
Linking .stack-work\dist\x86_64-windows\Cabal-1.22.2.0\build\dummy-project\dummy-project.exe ...
dummy-project-0.1.0.0: install
Installing executable(s) in
C:\temp1\dummy-project\.stack-work\install\x86_64-windows\nightly-2015-06-16\7.10.1\bin
```

No more warnings, except the first one that is not linked to stack but only to [GHC](https://ghc.haskell.org/trac/ghc/ticket/10454) I think.

I think it would be more *intuitive* to have `stack build` default to the host architecture (or the one in stack.yaml, if only one), even if there are concerns about GHC 64 bits on WIndows.

**Comment by snoyberg** [comment]
Cool, thanks for confirming. I think there's some lack of clarity still about how stack chooses architecture. In order of precedence:

* The `--arch` flag on the command line
* The `arch` config setting in stack.yaml
* The architecture that `stack` itself was built with

In other words, if you use a 64-bit `stack`, it will use a 64-bit GHC by default.

**Comment by wismill** [comment]
Ok, it makes sense. I suggest to add the 64-bit Windows `stack` to the [download page](https://github.com/commercialhaskell/stack/releases/latest).
